lib/mllp/client.ex

```diff
@@ -0,0 +1,444 @@
+  end
+
+  defp stop_connection(%State{} = state, error, context) do
+    if state.socket != nil do
+      telemetry(
+        :status,
+        %{status: :disconnected, error: format_error(error), context: context},
+        state
+      )
+
+      state.tcp.close(state.socket)
+    end
+
+    ensure_pending_reconnect_cancelled(state)
+  end
+
+  defp ensure_pending_reconnect_cancelled(%{pending_reconnect: nil} = state), do: state
+
+  defp ensure_pending_reconnect_cancelled(state) do
+    :ok = Process.cancel_timer(state.pending_reconnect, info: false)
+    %{state | pending_reconnect: nil}
+  end
+
+  defp attempt_connection(%State{} = state) do
+    telemetry(:status, %{status: :connecting}, state)
+    opts = [:binary, {:packet, 0}, {:active, false}] ++ state.socket_opts ++ state.tls_opts
+
+    case state.tcp.connect(state.address, state.port, opts, 2000) do
+      {:ok, socket} ->
+        state1 = ensure_pending_reconnect_cancelled(state)
+        telemetry(:status, %{status: :connected}, state1)
+        %{state1 | socket: socket, connect_failure: nil}
+
+      {:error, reason} ->
+        message = format_error(reason)
+        Logger.error(fn -> "Error connecting to #{state.socket_address} => #{message}" end)
+
+        telemetry(
+          :status,
+          %{status: :disconnected, error: format_error(reason), context: "connect failure"},
+          state
+        )
+
+        state
+        |> maintain_reconnect_timer()
+        |> Map.put(:connect_failure, reason)
+    end
+  end
+
+  defp maintain_reconnect_timer(state) do
+    ref = state.pending_reconnect || Process.send_after(self(), :timeout, 1000)
+    %State{state | pending_reconnect: ref}
+  end
+
+  defp maybe_receive_reply(state, message, %{expect_reply: true} = opts) do
+    case state.tcp.recv(state.socket, 0, opts.reply_timeout) do
+      {:ok, reply} ->
+        telemetry(:received, %{response: reply}, state)
+        maybe_verify_reply(state, message, reply, opts)
+
+      {:error, _} = err ->
+        err
+    end
+  end
+
+  defp maybe_receive_reply(_state, _msg, _opts) do
+    {:ok, :noop}
+  end
+
+  defp telemetry(_event_name, _measurements, %State{telemetry_module: nil} = _metadata) do
+    :ok
+  end
+
+  defp telemetry(event_name, measurements, %State{telemetry_module: telemetry_module} = metadata) do
+    telemetry_module.execute([:sender, event_name], add_timestamps(measurements), metadata)
+  end
+
+  defp add_timestamps(measurements) do
+    measurements
+    |> Map.put(:monotonic, :erlang.monotonic_time())
+    |> Map.put(:utc_datetime, DateTime.utc_now())
+  end
+
+  defp validate_options(opts) do
+    Map.get(opts, :address) || raise "No server address provided to connect to!"
+    Map.get(opts, :port) || raise "No server port provdided to connect to!"
+    opts
+  end
+
+  @default_opts %{
+    telemetry_module: MLLP.DefaultTelemetry,
+    tls_opts: [],
+    socket_opts: []
+  }
+
+  @default_send_opts %{
```

Comment:
We optimize for the case that the caller expect MLLP.Client to  wrap the message and expects the  reply to be verified.